transform/all_collection/spark/utils/common_df_transformations.py

```diff
@@ -0,0 +1,246 @@
+    fos_column = []
+
+    for subject in subjects:
+        fos = subject["subject"]["fos"]
+        fos_list = []
+        if fos:
+            for value in fos:
+                fos_list.append(value["value"])
+            fos_column.append(fos_list)
+        else:
+            fos_column.append(None)
+
+    harvested_properties[FOS] = fos_column
+
+
+def simplify_bestaccessright(df: DataFrame) -> DataFrame:
+    """Simplify best_access_right - get only label and convert structure to a string"""
+    return df.withColumn("bestaccessright", col("bestaccessright")["label"])
+
+
+def simplify_language(df: DataFrame) -> DataFrame:
+    """Simplify language - get only label and convert structure to a string"""
+    return df.withColumn("language", col("language")["label"])
+
+
+def create_open_access(
+    df: DataFrame, harvested_properties: Dict, col_name: str
+) -> None:
+    """Create boolean value whether record is open access or not, based on col_name"""
+    best_access_right = df.select(col_name).collect()
+    open_access_column = []
+
+    for access in best_access_right:
+        if access[col_name] in {
+            "OPEN",
+            "open access",
+            "fully_open_access",
+            "open_access",
+        }:
+            open_access_column.append(True)
+        else:
+            open_access_column.append(False)
+
+    harvested_properties[OPEN_ACCESS] = open_access_column
+
+
+def harvest_funder(df: DataFrame, harvested_properties: Dict) -> None:
+    """Harvest funder -> name and fundingStream as arr(arr(arr(fundingStream, <value>), arr(name, <value>>)))"""
+    projects_list = df.select("projects").collect()
+    funder_column = []
+
+    for projects in projects_list:
+        if projects["projects"]:
+            funder_list = []
+            for project in projects["projects"]:
+                funder = [
+                    ["fundingStream", project["funder"]["fundingStream"]],
+                    ["name", project["funder"]["name"]],
+                ]
+                funder_list.append(funder)
+            funder_column.append(funder_list)
+        else:
+            funder_column.append(None)
+
+    harvested_properties[FUNDER] = funder_column
+
+
+def harvest_url_and_document_type(df: DataFrame, harvested_properties: Dict) -> None:
+    """
+    Harvest url from instance.element.url as array(str)
+
+    Assumption:
+    - url has to be unique for specific record and it has to be a link
+    """
+    instances_list = df.select("instance").collect()
+    url_column = []
+    document_type_column = []
+
+    for instances in instances_list:
+        if instances["instance"]:
+            url_list = []
+            document_type_list = []
+            for instance in instances["instance"]:
+                if instance["url"]:
+                    for url in instance["url"]:
+                        if url and url not in url_list:
+                            url_list.append(url)
```

Comment:
I'm a person who likes fast failure and fast return, as also flat code. Did you consider writing it as such?

```
for instances in instances_list:
    if not instances["instance"]:
        continue

    url_list = []
    document_type_list = []
    for instance in instances["instance"]:
        if not instance["url"]:
            continue

        for url in instance["url"]:
            if not url or url in url_list:
                continue

            url_list.append(url)

# or in more map reduce form

non_empty_instances_list = [instance['instance'] for instance in instances_list if instance['instance']] 
for instances in non_empty_instances_list:
    url_list = []
    document_type_list = []
    instances_urls = [instance['url'] for instance in instances if instance['url']]
    for url in instances_urls:
        if not url or url in url_list:
            continue

        url_list.append(url)

Reply:
You are right that this nested code is ugly and could have been avoided. But the suggested code will now work properly,
by design creating dataframes and joining dataframes requires constant number of rows in columns so we can not continue in the loop, instead, we are supposed to append None or an empty array. Order matters. As a result, your suggested code after required changes will look more or less the same as my solution. For now I will keep it as it is. But I will refactor this later on